core/src/test/java/org/elasticsearch/cluster/action/shard/ShardFailedClusterStateTaskExecutorTests.java

```diff
@@ -0,0 +1,204 @@
+                throw new RuntimeException("simulated applyFailedShards failure");
+            }
+        };
+        List<ShardStateAction.ShardRoutingEntry> tasks = new ArrayList<>();
+        tasks.addAll(failingTasks);
+        tasks.addAll(nonExistentTasks);
+        ClusterStateTaskExecutor.BatchResult<ShardStateAction.ShardRoutingEntry> result = failingExecutor.execute(currentState, tasks);
+        Map<ShardStateAction.ShardRoutingEntry, Boolean> taskResultMap =
+            failingTasks.stream().collect(Collectors.toMap(Function.identity(), task -> false));
+        taskResultMap.putAll(nonExistentTasks.stream().collect(Collectors.toMap(Function.identity(), task -> true)));
+        assertTaskResults(taskResultMap, result, currentState, false);
+    }
+
+    private ClusterState createClusterStateWithStartedShards(String reason) {
+        int numberOfNodes = 1 + numberOfReplicas;
+        DiscoveryNodes.Builder nodes = DiscoveryNodes.builder();
+        IntStream.rangeClosed(1, numberOfNodes).mapToObj(node -> newNode("node" + node)).forEach(nodes::put);
+        ClusterState stateAfterAddingNode =
+            ClusterState.builder(clusterState).nodes(nodes).build();
+        RoutingTable afterReroute =
+            allocationService.reroute(stateAfterAddingNode, reason).routingTable();
+        ClusterState stateAfterReroute = ClusterState.builder(stateAfterAddingNode).routingTable(afterReroute).build();
+        RoutingNodes routingNodes = stateAfterReroute.getRoutingNodes();
+        RoutingTable afterStart =
+            allocationService.applyStartedShards(stateAfterReroute, routingNodes.shardsWithState(ShardRoutingState.INITIALIZING)).routingTable();
+        return ClusterState.builder(stateAfterReroute).routingTable(afterStart).build();
+    }
+
+    private List<ShardStateAction.ShardRoutingEntry> createExistingShards(ClusterState currentState, String reason) {
+        List<ShardRouting> shards = new ArrayList<>();
+        GroupShardsIterator shardGroups =
+            currentState.routingTable().allAssignedShardsGrouped(new String[] { INDEX }, true);
+        for (ShardIterator shardIt : shardGroups) {
+            for (ShardRouting shard : shardIt.asUnordered()) {
+                shards.add(shard);
+            }
+        }
+        List<ShardRouting> failures = randomSubsetOf(randomIntBetween(1, 1 + shards.size() / 4), shards.toArray(new ShardRouting[0]));
+        String indexUUID = metaData.index(INDEX).getIndexUUID();
+        int numberOfTasks = randomIntBetween(failures.size(), 2 * failures.size());
+        List<ShardRouting> shardsToFail = new ArrayList<>(numberOfTasks);
+        for (int i = 0; i < numberOfTasks; i++) {
+            shardsToFail.add(randomFrom(failures));
+        }
+        return toTasks(shardsToFail, indexUUID, reason);
+    }
+
+    private List<ShardStateAction.ShardRoutingEntry> createNonExistentShards(String reason) {
+        MetaData nonExistentMetaData =
```

Comment:
can we also add shards with the same index and id as the existing ones but with a *different* allocation ids?